pkg/controllers/at.go

```diff
@@ -0,0 +1,109 @@
+/*
+ (c) Copyright [2021] Micro Focus or one of its affiliates.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/names"
+	"github.com/vertica/vertica-kubernetes/pkg/paths"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// distributeAdmintoolsConf will copy admintools.conf to all of the running
+// pods.  It will target the base pod first since that is what will be used as
+// the basis for admintools.conf in subsequent iterations.  If copying to the
+// base pod is successful, then it will attempt to copy to all of the other pods
+// -- checking errors at the end to ensure we attempt at each pod.
+func distributeAdmintoolsConf(ctx context.Context, vdb *vapi.VerticaDB, vrec *VerticaDBReconciler,
+	pf *PodFacts, pr cmds.PodRunner, atConfTempFile string) error {
+	// We always distribute to a well known base pod first. The admintools.conf
+	// on this pod is used as the base for any subsequent changes.
+	basePod, err := findPodForFirstCopy(vdb, pf)
+	if err != nil {
+		return err
+	}
+	_, _, err = pr.CopyToPod(ctx, basePod, names.ServerContainer, atConfTempFile, paths.AdminToolsConf)
+	if err != nil {
+		return err
+	}
+
+	// Copy the admintools.conf to the rest of the pods.  We will do error
+	// checking at the end so that we try to copy it to each pod.
+	errs := []error{}
+	for _, p := range pf.Detail {
```

Comment:
The latest improvements of copying admintools.conf change looks good to me. Thanks!

Just some minor notes as potential future improvements: I see this is a sequential operation as it uses a for-loop, starting from v11, AT uses a new file copying mechanism, which improves the performance of file distribution to the entire cluster a lot. So, just in case that in the future people suggest/request updating this copying to be a parallel operation on a very large cluster, AT distribute_config_files could be used to distribute the files. At this time, using CopyToPod is better IMO.

Reply:
Okay, thanks.  This is good to know and I will keep it in mind.